File: sample_code/backend/python-fastapi/app/thales_protect_gateway_service.py
from cryptography import x509
from fastapi import APIRouter, Header

router = APIRouter()


class ThalesProtectGatewayService:
    def authenticateRequest(self, authorization: str | None) -> None:
        if not authorization:
            raise ValueError("Missing Authorization header")


@router.post("/protectInput")
def protectInput(payload: dict, authorization: str | None = Header(default=None)):
    ThalesProtectGatewayService().authenticateRequest(authorization)
    x509.load_pem_x509_certificate(payload["certificate"].encode("utf-8"))
    return {"route": "/protectInput", "status": "protected"}


@router.post("/protectInputAndCallLLM")
def protectInputAndCallLLM(payload: dict, authorization: str | None = Header(default=None)):
    ThalesProtectGatewayService().authenticateRequest(authorization)
    return {"route": "/protectInputAndCallLLM", "status": "protected-and-forwarded"}


@router.post("/revealInput")
def revealInput(payload: dict, authorization: str | None = Header(default=None)):
    ThalesProtectGatewayService().authenticateRequest(authorization)
    return {"route": "/revealInput", "status": "revealed"}

File: sample_code/backend/node-service/thales-protect-gateway-service.js
const express = require("express");
const jwt = require("jsonwebtoken");

class ThalesProtectGatewayService {
  constructor() {
    this.router = express.Router();
    this.router.post("/protectInput", this.authenticateRequest.bind(this), this.protectInput.bind(this));
    this.router.post("/protectInputAndCallLLM", this.authenticateRequest.bind(this), this.protectInputAndCallLLM.bind(this));
    this.router.post("/revealInput", this.authenticateRequest.bind(this), this.revealInput.bind(this));
  }

  authenticateRequest(req, res, next) {
    const token = req.headers.authorization || "";
    jwt.decode(token.replace("Bearer ", ""), { complete: true });
    next();
  }

  protectInput(req, res) {
    res.json({ route: "/protectInput", status: "protected" });
  }

  protectInputAndCallLLM(req, res) {
    res.json({ route: "/protectInputAndCallLLM", status: "protected-and-forwarded" });
  }

  revealInput(req, res) {
    res.json({ route: "/revealInput", status: "revealed" });
  }
}

module.exports = { ThalesProtectGatewayService };
